### PR TITLE
Use Fusion CB for splitting

### PR DESCRIPTION
**Pull request by dop-amin**
Extracted from #61 as introduced via d1a075f75b78c05671b012b5730d066a06935f7e.

### CODE REVIEW

File: slothy/targets/aarch64/aarch64_neon.py [hunk 1]
```diff
@@ -3175,7 +3176,50 @@ def core(inst,t,log=None):
 
     return core
 
-veor.global_fusion_cb  = eor3_fusion_cb()
+# TODO: Test only...
+# veor.global_fusion_cb  = eor3_fusion_cb()
+
+def eor3_splitting_cb():
```
Comment (mkannwischer) [inline slothy/targets/aarch64/aarch64_neon.py:3182]: Please add a comment here what this is doing. 
Probably also worth pointing out that we dont actually use this in any crypto examples so far.

File: slothy/core/slothy.py [hunk 2]
```diff
@@ -373,25 +374,12 @@ def fusion_loop(self, loop_lbl):
...
-        self.config.constraints.allow_reordering = False
-        self.config.sw_pipelining.enabled = False
-        self.config.split_heuristic = False
-        self.config.inputs_are_outputs = True
-        self.config.sw_pipelining.unknown_iteration_count = False
-        self.optimize_loop(loop_lbl)
-        self.config = c
-
-        assert SourceLine.is_source(self.source)
-
```
Comment (mkannwischer) [inline slothy/core/slothy.py:394]: Why was this removed?

File: slothy/core/slothy.py [hunk 3]
```diff
@@ -342,28 +342,29 @@ def _fusion_core(self, pre, body, post, logger):
...
 
-    def fusion_region(self, start, end):
+    def fusion_region(self, start, end, **kwargs):
         """Run fusion callbacks on straightline code"""
         logger = self.logger.getChild(f"ssa_{start}_{end}")
         pre, body, post = AsmHelper.extract(self.source, start, end)
 
         body_ssa = [ SourceLine(f"{start}:") ] +\
-             self._fusion_core(pre, body, logger) + \
+             self._fusion_core(pre, body, post, logger, **kwargs) + \
```
Comment (mkannwischer) [inline slothy/core/slothy.py:362]: I am a bit confused here. the argument `post` wasn't introduced in this PR. Did this just never work?

File: slothy/core/slothy.py [hunk 4]
```diff
@@ -342,28 +342,29 @@ def _fusion_core(self, pre, body, post, logger):
...
             [ SourceLine(f"{end}:") ]
         self.source = pre + body_ssa + post
         assert SourceLine.is_source(self.source)
 
-    def fusion_loop(self, loop_lbl):
+    def fusion_loop(self, loop_lbl, **kwargs):
```
Comment (mkannwischer) [inline slothy/core/slothy.py:367]: The functions `fusion_loop` and `fusion_region` are external APIs. 
Could you add documentation strings like we have for `optimize`, please? Does not have to be long.

**Review by mkannwischer — CHANGES_REQUESTED** [review]
Thanks @dop-amin!
Some some questions and suggestions.

File: slothy/core/slothy.py (hunk 2 above)
Comment (dop-amin) [reply slothy/core/slothy.py:394]: I think because it was never actually doing anything. In the `fusion_range`, it was never there in the first place. It makes no sense to call the actual optimizer in the fusion code, this is just supposed to perform replacements. 
But I think @hanno-becker made this change originally and may know better than I do.

File: slothy/core/slothy.py (hunk 3 above)
Comment (dop-amin) [reply slothy/core/slothy.py:372]: I think so, this was not used at all if I recall correctly. But again, @hanno-becker made this change originally and may know more.

File: slothy/targets/aarch64/aarch64_neon.py [hunk 5]
```diff
@@ -3175,7 +3176,50 @@ def core(inst,t,log=None):
 
     return core
 
-veor.global_fusion_cb  = eor3_fusion_cb()
+# TODO: Test only...
+# veor.global_fusion_cb  = eor3_fusion_cb()
```
[the remark this answers is not in the document]
Comment (dop-amin) [reply slothy/targets/aarch64/aarch64_neon.py]: I removed it and mention where we actually set the veor global_fusion_cb that these are two different options.

**Review by mkannwischer — APPROVED** [review]
Thanks for making the changes @dop-amin!
